mLRS/CommonTx/esp-jr_pin5_interface.h

```diff
@@ -0,0 +1,147 @@
+    void pin5_tx_enable(bool enable_flag);
+    virtual void parse_nextchar(uint8_t c) = 0;
+    virtual bool transmit_start(void) = 0; // returns true if transmission should be started
+
+    // actual isr functions
+    void uart_rx_callback(uint8_t c);
+    void uart_tc_callback(void);
+
+    // parser
+    typedef enum {
+        STATE_IDLE = 0,
+
+        // mBridge receive states
+        STATE_RECEIVE_MBRIDGE_STX2,
+        STATE_RECEIVE_MBRIDGE_LEN,
+        STATE_RECEIVE_MBRIDGE_SERIALPACKET,
+        STATE_RECEIVE_MBRIDGE_CHANNELPACKET,
+        STATE_RECEIVE_MBRIDGE_COMMANDPACKET,
+
+        // CRSF receive states
+        STATE_RECEIVE_CRSF_LEN,
+        STATE_RECEIVE_CRSF_PAYLOAD,
+        STATE_RECEIVE_CRSF_CRC,
+    } STATE_ENUM;
+#define STATE_TRANSMIT_START STATE_IDLE // Don't need this state for full duplex
+
+    // not used in this class, but required by the children, so just add them here
+    // no need for volatile since used only in isr context
+    uint8_t state;
+    uint8_t len;
+    uint8_t cnt;
+    uint16_t tlast_us;
+
+    // check and rescue
+    // the FRM303 can get stuck, whatever we tried, so brutal rescue
+    // can't hurt generally as safety net
+    uint32_t nottransmiting_tlast_ms;
+    void CheckAndRescue(void);
+};
+
+
+void tPin5BridgeBase::Init(void)
+{
+    state = STATE_IDLE;
+    len = 0;
+    cnt = 0;
+    tlast_us = 0;
+
+    telemetry_start_next_tick = false;
+    telemetry_state = 0;
+
+    nottransmiting_tlast_ms = 0;
+    uart_init();
+}
+
+
+void tPin5BridgeBase::TelemetryStart(void)
+{
+    telemetry_start_next_tick = true;
+}
+
+
+//-------------------------------------------------------
+// Interface to the uart hardware peripheral used for the bridge
+// called in isr context
+
+void tPin5BridgeBase::pin5_tx_enable(bool enable_flag)
+{
+    // nothing to do for full duplex
+}
+
+
+// we do not add a delay here before we transmit
+// the logic analyzer shows this gives a 30-35 us gap nevertheless, which is perfect
+
+void tPin5BridgeBase::uart_rx_callback(uint8_t c)
+{
+    // Not yet used
+    // parse_nextchar(c);
+}
+
+
+void tPin5BridgeBase::uart_tc_callback(void)
+{
+    // Not yet used
+    // pin5_tx_enable(false); // switches on rx
+    //state = STATE_IDLE;
+}
+
+
+//-------------------------------------------------------
+// Check and rescue
+// a good place to call it could be ChannelsUpdated()
+// Note: For the FRM303 it was observed that the TC callback may be missed in the uart isr, basically when
+// the jrpin5's uart isr priority is too low. This caused the jrpin5 loop to get stuck in STATE_TRANSMITING,
+// and not even channel data would be received anymore (= very catastrophic). This code avoids this.
+// With proper isr priorities, the issue is mainly gone, but the code remains, as safety net.
+//
+// For full duplex CRSF (internal module), this is used to poll for received bytes and pace telemetry
+
+void tPin5BridgeBase::CheckAndRescue(void)
```

Comment:
ok, this seems to be the key part

it seems you need to link into the code at this place and the call of this function appeared to be at the right place
not nice as you can guess, given the name of the function.
consider adding an extra function/method for what you need to do, with a proper name.

If CheckAndREscue is empty so it be. But this here doesn't really anything like that,

Reply:
Could we just rename it to something that covers both rescue and poll concepts?  They are a bit similar in function and I'd rather not introduce another.  How about CheckTransmit?

Reply:
I'm the opposite opinion. Better to disentangle. The two things do totally different things for the two mcus

Reply:
It really feels stupid to me to have two mutually exclusive entry points called from the same spot which both manage serial communication with the radio.  But I guess there is no point arguing.  So, should both be called and one is empty in each case?  Or should there be another ugly ifdef where they are called so we don't waste time calling an empty function?